Additional kubelet configuration settings

stmcginnis opened the pull request:
### Issue number:

Closes: #2877 
Closes: #1792 
Closes: #1445
Closes: #2977

### Description of changes:

This adds new settings that can be used for soft eviction, graceful shutdown, CPU quota enforcement, and memory manager policy.

The new settings are:

**Soft eviction**

- `settings.kubernetes.eviction-soft`
- `settings.kubernetes.eviction-soft-grace-period`
- `settings.kubernetes.eviction-max-pod-grace-period`

`eviction-soft` is the same as the existing `eviction-hard` but allows a grace period before pods are evicted. The `eviction-soft-grace-period` allows setting this grace period for each of the support eviction keys. `eviction-max-grace-period` is used to set an overall maximum for the eviction grace period.

**Graceful shutdown**

- `settings.kubernetes.shutdown-grace-period`
- `settings.kubernetes.shutdown-grace-period-for-critical-pods`

These allow setting a grace period to wait on node shut down to allow pods to exit. Of that grace period, a portion can be dedicated to making sure critical pods are shut down.

**CPU quota enforcement**

- `settings.kubernetes.cpu-cfs-quota-enforced`

This sets a flag telling kubelet whether or not to enforce quota for containers that specify CPU limits. The default behavior is to enforce these quotas.

For additional reference on the kubelet setting values, see [the node eviction documentation](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/) and the [`KubeletConfiguration` docs](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/#kubelet-config-k8s-io-v1beta1-KubeletConfiguration).

**Memory manager**

- `settings.kubernetes.memory-manager-policy`
- `settings.kubernetes.memory-manager-reserved-memory`

Allows configuring kubelet settings for the [memory manager](https://kubernetes.io/docs/tasks/administer-cluster/memory-manager/).

### Testing done:

Built and deployed a cluster. Configured these settings with:

```sh
apiclient apply << EOF
[settings.kubernetes.eviction-hard]
"memory.available" = "15%"

[settings.kubernetes.eviction-soft]
"memory.available" = "12%"

[settings.kubernetes.eviction-soft-grace-period]
"memory.available" = "30s"

[settings.kubernetes]
cpu-cfs-quota-enforced = false
shutdown-grace-period = "30s"
shutdown-grace-period-for-critical-pods = "10s"
eviction-max-pod-grace-period = 40
memory-manager-policy = "Static"

[settings.kubernetes.memory-manager-reserved-memory.0]
enabled = true
memory = "674Mi"
EOF
```

Verified values were accepted. The performed:

```sh
cat /etc/kubernetes/kubelet/config
```

Verified the config file had the expected configuration keys and values. Then ran:

```sh
systemctl status kubelet
journalctl -u kubelet
```

And made sure there were no errors or issues with service operation.

### Terms of contribution:

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

stmcginnis committed:
kubelet: Add `cpu-cfs-quota-enforced` setting

This adds a setting to be able to set the KubeletConfig value
`cpuCFSQuota` to `false`.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
kubelet: Add shutdown grace period settings

This adds the settings:

- `settings.kubernetes.shutdown-grace-period`
- `settings.kubernetes.shutdown-grace-period-for-critical-pods`

These control how long kubelet will wait on shutdown for pods to exit,
and what portion of that period to dedicate to waiting for critical pods
to exit.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis committed:
kubelet: Add soft eviction settings

This adds the settings:

- `settings.kubernetes.eviction-soft`
- `settings.kubernetes.eviction-soft-grace-period`
- `settings.kubernetes.eviction-max-pod-grace-period`

To configure the Kubelet soft eviction settings. This is similar to the
existing ``settings.kubernetes.eviction-hard` with additional
configuration needed to control the behavior of the soft eviction
thresholds.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

webern commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -485,27 +485,27 @@ enum EvictionSignal {
     }
 }
-string_impls_for!(KubernetesEvictionHardKey, "KubernetesEvictionHardKey");
+string_impls_for!(KubernetesEvictionKey, "KubernetesEvictionKey");
```
Can you convert these to use the new derive macro since they are being touched? You will no longer need a separate Enum and Struct, you can just have the enum.

See #2597, `sources/models/src/modeled_types/ecs.rs`

gthao313 commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -460,11 +460,11 @@ mod test_kubernetes_bootstrap_token {
 // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
 
-/// KubernetesEvictionHardKey represents a string that contains a valid Kubernetes eviction hard key.
+/// KubernetesEvictionKey represents a string that contains a valid Kubernetes eviction hard key.
```
Nit: should remove "hard" from the description as well?
```suggestion
/// KubernetesEvictionKey represents a string that contains a valid Kubernetes eviction key.
```

etungsten commented on sources/models/src/lib.rs:
```diff
@@ -228,10 +228,14 @@ struct KubernetesSettings {
+    eviction_hard: HashMap<KubernetesEvictionKey, KubernetesThresholdValue>,
+    eviction_soft: HashMap<KubernetesEvictionKey, KubernetesThresholdValue>,
+    eviction_soft_grace_period: HashMap<KubernetesEvictionKey, KubernetesDurationValue>,
+    eviction_max_grace_period: i32,
```
Do we need to constrain this value in anyway? Do we care about passing nonsensical values like `-100` to kubelet? We can use a modeled type to validate now that it supports scalar values.

stmcginnis replied:
I think the only constraint is that it is a non-negative value.

Do you have an example of how to set that constraint?

etungsten replied:
There is no existing setting that uses this yet but it is supported:
https://github.com/bottlerocket-os/bottlerocket/tree/develop/sources/models/scalar-derive#validating-a-number

gthao313 replied:
Maybe we can use regex to set the constraint? when I implemented eviction hard values, I used regex to limi the input value like this https://github.com/bottlerocket-os/bottlerocket/blob/develop/sources/models/src/modeled_types/kubernetes.rs#L538-L552

etungsten replied:
We want to avoid using strings for value types that are supposed to be integers. Users have already complained about this inconsistency: https://github.com/bottlerocket-os/bottlerocket/issues/2883

stmcginnis replied:
There are quite a few `i32` values in the kubernetes settings. I think nearly all of them have at least the constraint of needing to be non-negative. If we're going to enforce these constraints, we should probably do it for all of them.

Do we want to do that here, or would that be better kept it's only PR?

etungsten replied:
Let's do it for this now. Need to start somewhere. The other ones were added before `scalar-derive` was a thing. That's on my plate to resolve.

stmcginnis replied:
Error output isn't super great, but it does tell what the issue is:

```sh
[ssm-user@control]$ apiclient set --json '{"kubernetes": {"eviction-max-grace-period": -5}}'
Unable to deserialize input JSON into model: Unable to deserialize into PositiveInteger: number must be great than or equal to 0
[ssm-user@control]$ apiclient set --json '{"kubernetes": {"eviction-max-grace-period": 5}}'
[ssm-user@control]$ apiclient set --json '{"kubernetes": {"eviction-max-grace-period": 2147483647}}'
[ssm-user@control]$ apiclient set --json '{"kubernetes": {"eviction-max-grace-period": 2147483648}}'
Unable to deserialize input JSON into model: invalid value: integer `2147483648`, expected i32
```

stmcginnis committed:
settings: Enabling constraining i32 to non-negative value

This adds a `NonNegativeInteger` type to our model to be able to model
setting values that need to be between 0 and i32::MAX.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

webern commented on Release.toml:
```diff
@@ -192,4 +192,10 @@ version = "1.14.0"
 "(1.13.0, 1.13.1)" = [
     "migrate_v1.13.1_aws-profile-cred-provider.lz4",
 ]
+<<<<<<< HEAD
 "(1.13.1, 1.14.0)" = []
+=======
+"(1.13.1, 1.14.0)" = [
+    "migrate_v1.14.0_kubelet_config_settings.lz4",
+]
+>>>>>>> 2283ee47 (kubelet: Add `disable-cpu-cfs-enforcement` setting)
```
oops

stmcginnis replied:
I blame VS Code.

webern commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -1130,3 +1133,44 @@ mod test_valid_kmod_key {
+/// than an i32::MAX.
+#[derive(Debug, PartialEq, Scalar)]
+pub struct PositiveInteger {
+    inner: i32,
```
You might be able to make this work for `T: PartionOrd + SomethingAboutTheParse`, but maybe we don't have any `i16`, `i64`, or whatever.

stmcginnis replied:
I blame VS Code.

stmcginnis replied:
I thought about making it more flexible, but the problem with that is then the upper bounds could be too high.

There are some settings that take `i64` at least. But if I understand right, and please correct me if I'm wrong, if we made it more flexible it would make it possible to set an `i64` value for a field that eventually has to go into an `i32`. So I was thinking it would be safer to just be more verbose and have `PositiveInteger64`, `PositiveInteger16`, etc., as needed for whatever fields we want to contrain.

webern replied:
I would have to try it out to see if I could make it work with type `T`. No problem, it can be changed to generic later and the change would be an implementation detail.

jpculp commented:
Should https://github.com/bottlerocket-os/bottlerocket/issues/2742 be part of this as well?

stmcginnis commented:
> Should https://github.com/bottlerocket-os/bottlerocket/issues/2742 be part of this as well?

My initial reaction was yes, but after reading through that some more, I think it might be safer to keep that one separate. It has some special casing since it's only applicable to 1.25 and later.

I think if this one merges first, that one could just get added to the same migration if it ends up being in the same release cycle. But I think it would be better to keep that separate in its own PR.

kahirokunn commented:
I am very happy with this PR as I just needed a shutdown-grace-period !

stmcginnis commented:
Rebased to resolve merge conflicts.

stmcginnis committed:
kubelet: Add reserved memory settings

This enables the kubelet memory manager and allows setting its values
and the memory manager policy.

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

stmcginnis commented:
Added memory manager policy and settings.

bcressey commented on README.md:
```diff
@@ -456,50 +456,91 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.event-qps`: The maximum event creations per second.
 * `settings.kubernetes.eviction-hard`: The signals and thresholds that trigger pod eviction.
-  Remember to quote signals (since they all contain ".") and to quote all values.
+* `settings.kubernetes.eviction-max-grace-period`: Maximum grace period, in seconds, to wait for pod termination before soft eviction. Default is `0`.
```
Why not follow kubelet's name for the setting?

```suggestion
* `settings.kubernetes.eviction-max-pod-grace-period`: Maximum grace period, in seconds, to wait for pod termination before soft eviction. Default is `0`.
```

bcressey commented on README.md:
```diff
@@ -456,50 +456,91 @@ The following settings are optional and allow you to further configure your clus
 
   See the [Roles Anywhere documentation](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/credential-helper.html) for more details on the `aws_signing_helper` arguments.
 
+* `settings.kubernetes.disable-cpu-cfs-enforcement`: Whether to disable enforcing quotas for containers that specify CPU limits.
```
Typically we've preferred the positive form of this for settings, e.g. `enable` rather than `disable`.

It might be better to use a minor variation on the setting name, for clarity and discoverability. The goal would be to preserve the search phrase "cpu-cfs-quota" which aligns with the kubelet config:
```suggestion
* `settings.kubernetes.cpu-cfs-quota-enforced`: Whether CPU CFS quotas are enforced. Defaults to `true`.
```

stmcginnis replied:
I had a comment on this in the code, but I'm not seeing it now. I must have accidentally dropped it in a rebase. 

The problem that led me to make it `disable-*` is due to handlebars. Having `cpu-cfs-quota-enforced = false` is a problem since the default is that it is enforced. So needing to set it to false makes the templating a challenge with it being evaluated as "falsey".

I'll take another look, but I don't think this worked:

```
{{#if (eq settings.kubernetes.cpu-cfg-quota-enforced false)}}
enforced = false
{{/if}}
```

I'll take another shot and see if I can find a better way.

stmcginnis replied:
Figured it out - brain got stuck on the wrong pattern. :]

bcressey commented on README.md:
````diff
@@ -456,50 +456,91 @@ The following settings are optional and allow you to further configure your clus
   ```
 
+  Remember to quote signals (since they all contain ".") and to quote all values.
+
````
I find this sentence confusing now, since it's further away from the context of what signals are.

bcressey commented on README.md:
```diff
@@ -456,50 +456,91 @@ The following settings are optional and allow you to further configure your clus
 * `settings.kubernetes.kube-api-qps`: The QPS to use while talking with kubernetes apiserver.
 * `settings.kubernetes.log-level`: Adjust the logging verbosity of the `kubelet` process.
   The default log level is 2, with higher numbers enabling more verbose logging.
+* `settings.kubernetes.memory-manager-policy`: The memory management policy to use: `None` (default) or `Static`.
```
Let's move the documentation down closer to `reserved-memory` to group related settings together.

For alignment with `cpu-manager-policy`, I'd also prefer to document "none" and "static" as the accepted values, even if we accept and emit the upper-cased form in code + templates.

stmcginnis replied:
I double checked, but while `cpu-manager-policy` takes `none` and `static` (must be lower case), `memory-manager-policy` takes `None` and `Static` (must be title cased). So unfortunately there's an inconsistency, but it's an upstream inconsistency. Kubelet will fail to start if either of the settings are not cased correctly.

stmcginnis replied:
> Let's move the documentation down closer to reserved-memory to group related settings together.

All current settings are in alphabetical order. I'd kind of hate to break that for one setting. What if I just add a "see `reserved-memory` for static memory manager configuration"?

stmcginnis replied:
> What if I just add...

Oh, I did that with the next line. Let me know if that's not enough. For now I will also add a note to `reserved-memory` to reference back to this setting.

stmcginnis replied:
One other idea if you don't think the cross referencing is enough. It might actually be nice to prepend `memory-manager-reserved-memory`. That would allow us to keep alphabetical sorting while keep these related settings together. It would also make it clear that the `memory-manager-reserved-memory` settings are related to the `memory-manager-policy`.

Just let me know what you think.

bcressey replied:
> It might actually be nice to prepend memory-manager-reserved-memory. That would allow us to keep alphabetical sorting while keep these related settings together.

Yes, please - I like this idea.

bcressey replied:
> I double checked, but while cpu-manager-policy takes none and static (must be lower case), memory-manager-policy takes None and Static (must be title cased). So unfortunately there's an inconsistency, but it's an upstream inconsistency. Kubelet will fail to start if either of the settings are not cased correctly.

We should be able to use [serde aliases](https://serde.rs/field-attrs.html#alias) to paper over the inconsistency:
```
 #[derive(Debug, Clone, Eq, PartialEq, Hash, Serialize, Deserialize, Scalar)]
 pub enum KubernetesMemoryManagerPolicy {
+    #[serde(alias = "static")]
     Static,
+    #[serde(alias = "none")]
     None,
 }
```

And perhaps give the CPU manager policy struct the same treatment. Then they will accept both forms (`"Static"` and `"static"`) when deserializing, but the expected form will be used for all serialization.

stmcginnis replied:
`ValidCpuManagerPolicy` has `#[serde(rename_all = "lowercase")]` - that would be functionally equivalent, right?

bcressey replied:
> `ValidCpuManagerPolicy` has `#[serde(rename_all = "lowercase")]` - that would be functionally equivalent, right?

That doesn't quite do the same thing - it means serde will expect the lowercase form when deserializing, and serialize into the lowercase form. Which is correct for the CPU manager policy, but not the memory manager policy.

What we want is to always serialize / deserialize into the expected form, but *also* to accept certain aliases when deserializing. So we'd need to add `Static` and `None` aliases on `ValidCpuManagerPolicy` variants (it already accepts / expects "static" and "none").

stmcginnis replied:
Done!

bcressey commented on README.md:
```diff
@@ -456,50 +456,91 @@ The following settings are optional and allow you to further configure your clus
+
 * `settings.kubernetes.server-tls-bootstrap`: Enables or disables server certificate bootstrap. When enabled, the kubelet will request a certificate from the certificates.k8s.io API. This requires an approver to approve the certificate signing requests (CSR). Defaults to `true`.
+* `settings.kubernetes.shutdown-grace-period`: Delay the node should wait for pod termination before shutdown. Default is `0s`.
+* `settings.kubernetes.shutdown-critical-pod-grace-period`: The portion of the shutdown delay that should be dedicated to critical pod shutdown. Default is `0s`.
```
Maybe one of these instead? The goal would be to avoid breaking up the search phrase "shutdown grace period".

```suggestion
* `settings.kubernetes.shutdown-grace-period-for-critical-pods`: The portion of the shutdown delay that should be dedicated to critical pod shutdown. Default is `0s`.
```

```suggestion
* `settings.kubernetes.critical-pod-shutdown-grace-period`: The portion of the shutdown delay that should be dedicated to critical pod shutdown. Default is `0s`.
```

(I have a slight preference for the first.)

bcressey commented on README.md:
```diff
@@ -456,50 +456,91 @@ The following settings are optional and allow you to further configure your clus
+  [settings.kubernetes]
+  "memory-manager-policy" = "Static"
+
+  [settings.kubernetes.reserved-memory.0]
+  "enabled" = true
+  "memory" = "674Mi"
+  "hugepages-1Gi" = "1Gi"
+
+  [settings.kubernetes.reserved-memory.1]
+  "enabled" = true
+  "hugepages-2Mi" = "2148Mi"
```
Am I reading this correctly?
* first example - 1 GiB memory reserved in the form of a single 1 GiB huge page, in addition to the 674 MiB reservation
* second example - 2148 MiB memory reserved in the form of 1,074 2 MiB huge pages

If not, it might be good to add comments inline on what these settings achieve.

bcressey commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -1324,3 +1307,98 @@ pub struct CredentialProvider {
+enum MemoryReservationResources {
+    #[serde(rename = "memory")]
+    Memory,
+    #[serde(rename = "hugepages-2Mi")]
+    HugePages2Mi,
+    #[serde(rename = "hugepages-1Gi")]
+    HugePages1Gi,
```
Can you confirm that `hugepages-2Mi` and `hugepages-1Gi` are not mutually exclusive? I was originally under the impression that they might be, but from the docs I've drawn the impression that these are treated as three distinct types of memory resource limits.

stmcginnis replied:
The docs aren't great here, but these are treated as distinct types of memory resource limits. The user would need to determine what is available by using `kubectl describe node` to see what is reported as the available values:

```
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                    Requests    Limits
  --------                    --------    ------
  cpu                         325m (16%)  0 (0%)
  memory                      140Mi (2%)  340Mi (5%)
  ephemeral-storage           0 (0%)      0 (0%)
  hugepages-1Gi               0 (0%)      0 (0%)
  hugepages-2Mi               0 (0%)      0 (0%)
```

These would vary based on the type of hardware and/or the instance type.

stmcginnis replied:
@foersleo or @markusboehme - can you tell if anything else is needed to [support using hugepages](https://www.kernel.org/doc/Documentation/vm/hugetlbpage.txt)?

From what I've read, this would require passing in kernel parameters on boot, but I'm not clear on the full implications of that and how things are handled with Bottlerocket.

markusboehme replied:
You're right, some kernel configuration is required. At some point the kernel needs to be told how many huge pages of which size to set aside. This does not necessarily have to happen at boot, it can also be done at any later point in time via procfs or sysfs. However, a huge page is a long stretch of physically contiguous memory, and physical memory will get fragmented over the course of normal system operation. This means the chance of successfully allocating huge pages decreases over time. Hence, it's best to do so early.

In Bottlerocket, the above translates to setting the kernel command line via the user data mechanism or setting 
corresponding values in sysfs. For example, to set aside 1024x 2 MiB huge pages at boot, one might use

```toml
[settings.boot]
reboot-to-reconcile = true

[settings.boot.kernel]
hugepages = ["1024"]
```

To not pay the price of a reboot (modifying the kernel command line happens after first boot), an alternative would be to set a sysctl (this happens later in the boot, though, after Bottlerocket's corndog is running):

```toml
[settings.kernel.sysctl]
"vm.nr_hugepages" = "1024"
```

Or with a bootstrap container writing `/sys/kernel/mm/hugepages/hugepages-*/nr_hugepages`, here with control over which size huge pages to set aside.

Some implications/things to watch out for:

* Huge pages set aside this way cannot be used for other, regular allocations. Corollaries: 1. If no pods use them, they're dead memory. 2. They add to memory pressure. Workloads can run out of memory earlier if they're not appropriately sized ("requests" in pod spec).
* Huge pages cannot be swapped out. Not a problem for Kubernetes, since swap is not supported anyway.
* Requesting too many huge pages at boot time (i.e. close to or more than the system memory) prevents the system from booting.

bcressey commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -1324,3 +1307,98 @@ pub struct CredentialProvider {
+/// KubernetesMemoryReservationKey represents a string that contains a valid Kubernetes memory
+/// resource reservation key.
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct KubernetesMemoryReservationKey {
+    inner: String,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+enum MemoryReservationResources {
```
Why do we need the "inner" pattern here but we were able to refactor `KubernetesEvictionHardKey` /  `EvictionSignal` into the new `KubernetesEvictionKey` enum?

webern replied:
You are right, the struct can go away. See better patterns here:
- https://github.com/bottlerocket-os/bottlerocket/pull/2597
- https://github.com/bottlerocket-os/bottlerocket/pull/2697

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -1130,3 +1133,44 @@ mod test_valid_kmod_key {
+    use super::PositiveInteger;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn valid_kmod_key() {
+        assert!(PositiveInteger::try_from(0).is_ok());
+        assert!(PositiveInteger::try_from(i32::MAX).is_ok());
+        assert!(PositiveInteger::try_from(42).is_ok());
+    }
+
+    #[test]
+    fn invalid_kmod_key() {
+        assert!(PositiveInteger::try_from(i32::MIN).is_err());
+        assert!(PositiveInteger::try_from(-1).is_err());
+    }
```
```suggestion
    #[test]
    fn valid_positive_integer() {
        assert!(PositiveInteger::try_from(0).is_ok());
        assert!(PositiveInteger::try_from(i32::MAX).is_ok());
        assert!(PositiveInteger::try_from(42).is_ok());
    }

    #[test]
    fn invalid_positive_integer() {
        assert!(PositiveInteger::try_from(i32::MIN).is_err());
        assert!(PositiveInteger::try_from(-1).is_err());
    }
```

webern commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -1324,3 +1307,98 @@ pub struct CredentialProvider {
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// KubernetesMemoryReservationKey represents a string that contains a valid Kubernetes memory
+/// resource reservation key.
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct KubernetesMemoryReservationKey {
+    inner: String,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+enum MemoryReservationResources {
+    #[serde(rename = "memory")]
+    Memory,
+    #[serde(rename = "hugepages-2Mi")]
+    HugePages2Mi,
+    #[serde(rename = "hugepages-1Gi")]
+    HugePages1Gi,
+}
```
```suggestion
/// KubernetesMemoryReservationKey represents a string that contains a valid Kubernetes memory
/// resource reservation key.
#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Scalar)]
enum KubernetesMemoryReservationKey {
    #[serde(rename = "memory")]
    Memory,
    #[serde(rename = "hugepages-2Mi")]
    HugePages2Mi,
    #[serde(rename = "hugepages-1Gi")]
    HugePages1Gi,
}
```

Again for better rendering:

```rust
/// KubernetesMemoryReservationKey represents a string that contains a valid Kubernetes memory
/// resource reservation key.
#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize, Scalar)]
enum KubernetesMemoryReservationKey {
    #[serde(rename = "memory")]
    Memory,
    #[serde(rename = "hugepages-2Mi")]
    HugePages2Mi,
    #[serde(rename = "hugepages-1Gi")]
    HugePages1Gi,
}
```

stmcginnis commented:
Rebased to address conflict in `Release.toml` due to 1.13.3 and 1.13.4.

zmrow reviewed (approved):
LGTM!

bcressey commented on sources/api/migration/migrations/v1.14.0/kubelet-config-settings/src/main.rs:
```diff
@@ -5,6 +5,14 @@ use std::process;
+        "settings.kubernetes.eviction-soft-grace-period",
+        "settings.kubernetes.eviction-max-pod-grace-period",
+        "settings.kubernetes.memory-manager-policy",
+        "settings.kubernetes.memory-manager-reserved-memory",
```
This one needs to go in a different `AddPrefixes` migration since we allow user-defined keys underneath it.

stmcginnis replied:
Added/moved.

bcressey commented on sources/models/src/modeled_types/kubernetes.rs:
```diff
@@ -1358,3 +1339,75 @@ mod test_kubernetes_cpu_manager_policy_option {
+
+    #[test]
+    fn good_policy_key() {
+        for ok in &["Static", "None"] {
```
```suggestion
        for ok in &["Static", "static", "None", "none"] {
```

stmcginnis replied:
Done

bcressey commented on sources/models/src/lib.rs:
```diff
@@ -255,6 +263,10 @@ struct KubernetesSettings {
     server_key: ValidBase64,
+    shutdown_grace_period: KubernetesDurationValue,
+    shutdown_grace_period_for_critical_pods: KubernetesDurationValue,
+    memory_manager_reserved_memory: HashMap<Identifier, KubernetesMemoryReservation>,
```
Is `Identifier` the best type for the key or should it be another `PositiveInteger`?

stmcginnis replied:
It does end up just being a numeric identifier. But I don't think we can enforce that here since integer values can't be used as HashMap keys. It is an "Identifier" to kubelet, so it's also probably better to have kubelet ignore an invalid setting than to fail boot if someone has something strange set up in their user data.

bcressey commented on sources/models/src/modeled_types/shared.rs:
```diff
@@ -1130,3 +1133,44 @@ mod test_valid_kmod_key {
+/// Input value that needs to be a positive value, but should not be greater
+/// than an i32::MAX.
+#[derive(Debug, PartialEq, Scalar)]
+pub struct PositiveInteger {
```
Should this be named `NonNegativeInteger` since zero is allowed and is neither positive or negative?

I realize that's a bit of a math nitpick but I could see someone assuming that the PositiveInteger type would exclude zero without looking too closely at the implementation.

stmcginnis replied:
Renamed to appease all mathematical purists. :D

stmcginnis committed:
kubelet: Add aliases for ValidCpuManagerPolicy

Signed-off-by: Sean McGinnis <stmcg@amazon.com>

sanjeevpandey19 commented:
Hi @stmcginnis, First of all, thanks for this PR. Just out of curiosity, wanted to know if this is going to be part of the 1.14.0 release.

stmcginnis commented:
> Hi @stmcginnis, First of all, thanks for this PR. Just out of curiosity, wanted to know if this is going to be part of the 1.14.0 release.

Hi @sanjeevpandey19! Yes, hoping to have this part of the 1.14.0 release. Everything is subject to change based on what is found and testing and other issues of course, but hoping to have this in a 1.14.0 release in the very near future. ;)
